pkg/codegen/hcl2/model/type.go

```diff
@@ -43,13 +45,84 @@ type Type interface {
 	ConversionFrom(src Type) ConversionKind
 	String() string
 
+	hash(stack objTypeSet) uint32
 	equals(other Type, seen map[Type]struct{}) bool
 	conversionFrom(src Type, unifying bool, seen map[Type]struct{}) (ConversionKind, lazyDiagnostics)
 	string(seen map[Type]struct{}) string
 	unify(other Type) (Type, ConversionKind)
 	isType()
 }
 
+type objTypeSet map[*ObjectType]struct{}
+
+func (s *objTypeSet) add(t *ObjectType) bool {
```

Comment:
Something just slightly weird here, I'm more used to set.Add to return true if the element is added, here it's inverse, it returns false. E.g. https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.add?view=net-5.0